### PR TITLE
(PCP-874) Add functionality for recursive downloads in download_file

### PR DESCRIPTION
**Pull request by McdonaldSeanp**
This work should expand the download_file module in pxp-agent to implement taking an array of files/directories/symlinks and creating them all on the system.

The schema for download_file already takes an array (previously it just operated on the first element), so the only update required for the schema is a field to differentiate between files/dirs/symlinks.

### CODE REVIEW

**Comment by McdonaldSeanp** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2012r2-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2012r2-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2012r2-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2012r2-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2012r2-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2012r2-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2012r2-64a

**Comment by McdonaldSeanp** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2012r2-64a

File: lib/tests/unit/modules/download_file_test.cc [hunk 1]
```diff
@@ -191,6 +222,16 @@ TEST_CASE("Modules::DownloadFile::callAction", "[modules]") {
         REQUIRE(fs::is_regular_file(fs::path(TEST_FILE_DIR + "/file.txt")));
     }
 
+    SECTION("Correctly returns exitcode 1 if the new symlink requested is already a file") {
+        // FAILURE_DIR_EXISTS_AS_FILE_CONTENT will fail because it will make a request to make a
```
Comment (speedofdark) [inline lib/tests/unit/modules/download_file_test.cc]: comment doesn't match code at line 228.

**Comment by McdonaldSeanp** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2012r2-64a

File: lib/src/modules/download_file.cc [hunk 2]
```diff
@@ -113,31 +119,71 @@ namespace Modules {
...
-    } catch (Module::ProcessingError& e) {
-      LOG_ERROR("Failed to download {1}; {2}", destination, e.what());
-      response.output = write_download_results(results_dir, EXIT_FAILURE, "", lth_loc::format("Failed to download {1}; {2}", destination, e.what()));
+    for (auto this_file : files) {
+      auto destination = fs::path(this_file.get<std::string>("destination"));
+      auto file_type = this_file.get<std::string>("file_type");
+      if (file_type == "file") {
+        try {
+          if (!fs::exists(destination.parent_path())) {
+            Util::createDir(destination.parent_path());
```
Comment (branan) [inline lib/src/modules/download_file.cc]: Why is this necessary?

File: lib/src/modules/download_file.cc [hunk 3]
```diff
@@ -113,31 +119,71 @@ namespace Modules {
...
+            return failure_response(request, results_dir, lth_loc::format("Destination {1} already exists and is not a directory!", destination));
+          }
+        } else {
+          try {
+            Util::createDir(destination);
+          } catch (fs::filesystem_error& e) {
+            return failure_response(request, results_dir, lth_loc::format("Failed to create directory {1}; {2}", destination, e.what()));
+          }
+        }
+      } else if (file_type == "symlink") {
```
Comment (branan) [inline lib/src/modules/download_file.cc]: Have we confirmed how Bolt and Puppet behave for symlinks (on POSIX and Windows) so we can be consistent?

File: lib/src/util/bolt_helpers.cc [hunk 4]
```diff
@@ -188,6 +188,11 @@ namespace Util {
     fs::permissions(dir, NIX_DIR_PERMS);
   }
 
+  void createSymLink(const fs::path& destination, const fs::path& source) {
+    fs::create_symlink(destination, source);
+    fs::permissions(destination, NIX_DIR_PERMS);
```
Comment (branan) [inline lib/src/util/bolt_helpers.cc]: This is a race. In theory someone could access the symlink with default permissions in between creation and the permissions fix. We should be able to specify the permissions at creation time (on POSIX with UMASK, on Windows just as part of the CreateFile IIRC). Dunno how much of a refactor there is to achieve that...

File: lib/src/modules/download_file.cc (hunk 3 above)
Comment (McdonaldSeanp) [reply lib/src/modules/download_file.cc]: There's an acceptance test that uses puppet to check that puppet is reporting the symlink as:

`ensure => link`

So I _think_ they behave the same.

File: lib/src/modules/download_file.cc (hunk 2 above)
Comment (McdonaldSeanp) [reply lib/src/modules/download_file.cc]: Now that I think about it I think this is in the wrong place. downloadFileFromMaster doesn't create the directory itself, it downloads the file to a cachedir and renames it (using fs::rename()) if the SHA matches. So if the underlying destination dir for the file doesn't exist yet it will break.

Though after looking at it I think downloadFileFromMaster _should_ be creating the underlying destination dir if it doesn't exist, so that it can do so only after the matching SHA is determined.

File: lib/src/util/bolt_helpers.cc (hunk 4 above)
Comment (McdonaldSeanp) [reply lib/src/util/bolt_helpers.cc]: So after some digging on this:

It looks like there are ways for POSIX nodes to specify umask of the pxp-agent service, which means any file/symlink/dir created with boost should match the umask of the service process. That means between the file being created and the call to fs::permissions it looks like the user can still control the security of the files/symlinks/dirs.

For windows: all permissions are inherited by the containing directory, meaning there is no scenario where files/symlinks will have perms more open than the containing directories.

Since download_file isn't supporting specialized permissions from the user I believe the combination of those things means we don't need to worry about the race.

**Comment by McdonaldSeanp** [comment]
jenkins please test this on redhat7-64a,ubuntu1804-64a,windows2012r2-64a

File: acceptance/tests/pxp_module_bolt_download_file/download_file.rb [hunk 5]
```diff
@@ -63,33 +68,74 @@ def test_file_does_not_exist(agent)
     on master, puppet('resource service puppetserver ensure=running')
   end
 
-  step 'execute successful download_file' do
-    clean_files(suts)
+  step 'execute successful download_file with files and directories' do
```
Comment (speedofdark) [inline acceptance/tests/pxp_module_bolt_download_file/download_file.rb]: files, links, directories? Or is that too pedantic?

File: acceptance/tests/pxp_module_bolt_download_file/download_file.rb [hunk 6]
```diff
@@ -63,33 +68,74 @@ def test_file_does_not_exist(agent)
...
+      request << download_file_entry('', '', test_files[0], test_symlink, 'symlink')
+      run_successful_download(master,
+                              agent,
+                              request)
+      test_files.each do |file|
+        test_file_resource_exists(agent, file, 'file')
+      end
+      test_file_resource_exists(agent, test_dir, 'directory')
+      test_file_resource_exists(agent, test_symlink, 'link')
+      teardown {
```
Comment (speedofdark) [inline acceptance/tests/pxp_module_bolt_download_file/download_file.rb]: does something else take care of cleaning non-file resources?

File: lib/tests/unit/modules/download_file_test.cc [hunk 7]
```diff
@@ -58,17 +60,68 @@ static const std::string CRT { "mock_crt" };
...
+                                                                    "},"
+                                                                    "\"sha256\":\"\","
+                                                                    "\"destination\":\"%1%/file.txt\","
+                                                                    "\"link_source\":\"\","
+                                                                    "\"file_type\":\"directory\""
+                                                                "}"
+                                                            "]"
+                                                       "}") % TEST_FILE_DIR;
+
+// Create a failure scenario by using a request with a "directory" file type where
```
Comment (speedofdark) [inline lib/tests/unit/modules/download_file_test.cc]: directory file type?

Comment (McdonaldSeanp) [reply lib/tests/unit/modules/download_file_test.cc]: Yeah, the field in the schema I used to identify if it was a file/dir/link is called "file_type". I can change that.

File: acceptance/tests/pxp_module_bolt_download_file/download_file.rb (hunk 6 above)
Comment (McdonaldSeanp) [reply acceptance/tests/pxp_module_bolt_download_file/download_file.rb]: Oh, whoops, no.

File: lib/src/modules/download_file.cc [hunk 8]
```diff
@@ -113,31 +119,68 @@ namespace Modules {
...
-                    destination,
-                    files[0]);
-      response.output = write_download_results(results_dir, EXIT_SUCCESS, "downloaded", "");
-    } catch (Module::ProcessingError& e) {
-      LOG_ERROR("Failed to download {1}; {2}", destination, e.what());
-      response.output = write_download_results(results_dir, EXIT_FAILURE, "", lth_loc::format("Failed to download {1}; {2}", destination, e.what()));
+    for (auto this_file : files) {
+      auto destination = fs::path(this_file.get<std::string>("destination"));
+      auto content_type = this_file.get<std::string>("content_type");
+      if (content_type == "file") {
```
Comment (caseywilliams) [inline lib/src/modules/download_file.cc]: `content_type` seems a little strange as the name of the parameter, because it makes me think about the actual contents of the file/symlink/directory rather than the type of thing it is - I think `file_type` or `kind` or something like that might work.

But do we have to add a type param at all? I'm wondering if we could use a symlink if the `link_source` param is set, and otherwise use the same strategy as before?

(If this is here and called content_type to match some other code somewhere else, ignore me)

Comment (caseywilliams) [reply lib/src/modules/download_file.cc]: Oh, nevermind about the second part, I was looking at the changeset as a whole, I didn't realize we kind of need the new parameter to deal with directories

File: lib/src/util/bolt_helpers.cc (hunk 4 above)
Comment (caseywilliams) [reply lib/src/util/bolt_helpers.cc]: Can you leave a comment near this code with a summary of that info for later?

File: lib/src/modules/download_file.cc (hunk 8 above)
Comment (McdonaldSeanp) [reply lib/src/modules/download_file.cc]: I had this var named "file_type" before, but Jonathan pointed out that it made things confusing to talk about (i.e. "file_type":"file" is awkward language to use).

I'll change it to `kind` though. I think that's probably a good idea.
